### PR TITLE
Fixes a DI context issue between CLI and server injection.

### PR DESCRIPTION
**Pull request by nyarly**
There was a problem with how DI contexts were being passed around: the command object responsible for `sous server` registered providers with its DI, but the actual server component didn't get that DI - it got one from further up the chain of contexts, which meant that important machinery wasn't available - so the server would blow up when certain requests were made of it.

Fixing this involved pulling a fair amount of the plumbing out and packing it back in.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
LGTM

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: cli/cli.go [hunk 1]
```diff
@@ -91,11 +93,26 @@ func BuildCLIGraph(cli *CLI, root *Sous, in io.Reader, out, err io.Writer) *grap
...
 
 	return cli.SousGraph
 }
 
+func (cli *CLI) scopedGraph(cmd, under cmdr.Command) *graph.SousGraph {
+	if g, ok := cli.scopedGraphs[cmd]; ok {
+		return g
+	}
+
+	parent := cli.scopedGraphs[under]
```
Comment (SegFaultAX) [inline cli/cli.go]: Does this have the potential to cause a panic?

File: cli/cli.go [hunk 2]
```diff
@@ -91,11 +93,26 @@ func BuildCLIGraph(cli *CLI, root *Sous, in io.Reader, out, err io.Writer) *grap
...
+
+	g := &graph.SousGraph{Psyringe: parent.Clone()}
+	if r, ok := cmd.(Registrant); ok {
```
Comment (SegFaultAX) [inline cli/cli.go]: I haven't seen this form yet `cmd.(Registrant)`, what's the significance of it?

File: cli/cli.go (hunk 1 above)
Comment (nyarly) [inline cli/cli.go]: It does, but under normal operations it shouldn't - only if you pass in the non-parent of a missing command. (At one point a test caused panics.) If that happens, there's not much to do *but* panic.

File: cli/cli.go (hunk 2 above)
Comment (nyarly) [inline cli/cli.go]: It's a type assertion: "cmd fulfills the Registrant interface" in this case. The two-arg result works like map access: ok will be true if the assertion holds. (There's a single-arg form that panics if the assertion fails.)

Comment (SegFaultAX) [inline cli/cli.go]: I see! Interesting syntax for that. Thanks!

File: cli/cli.go (hunk 1 above)
Comment (SegFaultAX) [inline cli/cli.go]: Cool, thanks.

**Pull request merged**